### ISSUE TITLE
[KED-1727] Only trigger CI/CD (CircleCI) when there are code related changes in the repo

### ISSUE THREAD

**Issue by neomatrix369** [quantumblacklabs/kedro#269]
## Description

Currently, every commit results in a CI/CD trigger even when non-code related changes are made in the pull request. Especially when a lot of docs or README changes are made.

This will help release build resources and also not put the PR on hold till the build process is finished, especially helpful when small changes are made and many of them have made them - reducing the number of PRs waiting to be approved and/or merged.

## Context

We want to limit this to ONLY build when code related changes are made, any file changes that directly or indirectly results in changes to the library/package. In additional files that are related to building the documentation i.e. Sphinx related. And also configuration changes that are related to code and/or docs building.

For all other changes i.e. shell scripts or things that do not impact the build at all these should result in skipping the build.

Or there should be a leaner build process for other changes in case formatting or anything else needs to be checked. 

## Possible Implementation

Either find a way to make CircleCI detect such changes by creating some hook/call back to a script which then returns a trigger asking CircleCI to "go ahead and build" or "not to go build" - not sure if this is available as part of CircleCI's services/APIs.

Another way would be to go through the list of files that have been changed, and if a folder or file is flagged as source code or docs (sphinx) then trigger build or else not. 

_It's possible I have not enlisted all classes of files/folders when changed should result in triggering a build or not, please add comments below stating them._

## Possible Alternatives

CircleCI like many other CI services allow adding the `[skip ci]` to your commit message(s) and this tells the CI service that do not build this commit, it not code related. For e.g. please see the commit message on 14ba92314eb2fe6f7580b11c7fcd8a45fc477421.

Although this one can be prone to false positives or mistakes by the developer, and will need to be checked during code review process and initiate a build action on the PR.

**Comment by neomatrix369** [comment]
Is this a desirable thing to do wrt to the CI/CD build system for Kedro? Happy to discuss this further.

**Comment by mzjp2** [comment]
> Is this a desirable thing to do wrt to the CI/CD build system for Kedro? Happy to discuss this further.

Thanks for raising this. IMO, I don't think the added complexity here is worth the small amount of time saved. We run all our build steps in parallel, so they all complete within ~5-8 minutes, removing unit tests for a pure docs build PR will still take ~5-8 minutes because it runs the docs, but doesn't run unit tests (which would have run in parallel anyway).

Moreover, it's hard to make sure we've run everything required, e.g updating code might update the docstrings which should trigger a doc build, but how do you check if a doctring has been updated in a simple way?

Happy to hear contrary viewpoints though! 😀

**Comment by neomatrix369** [comment]
@mzjp2 Sorry for the delay in reply, so good points raised although my views are:
- it's about separating the wheat from the chaff (if that's the right terminology)
- always trying to reflect the actual as much as possible (less to check)
- I don't see much complexity here - if the separation between included files/folders (or excluded files/folders) is clear

What do I mean by that?
- atm you will agree there might be false positives (or false negatives) on pull requests that seem to no comply with your build requirements for ones that do not make code or documentation changes (that require building), an example is this PR https://github.com/quantumblacklabs/kedro/pull/268 - it could show a green tick but it's not clear which item(s) it does not comply to
- these could delay your approving and merging and you have to manually check for them sometimes, same for the person who has raised the PR
- every commit triggers a build, even if it's ~5 minutes: many commits on many pull requests can make users wait till they know their PR/commit is good to go, plus the above thing mentioned

These may look small issues but as more contributors chime in you will see this coming in the way for both parties - and more communications happen (atm its low latency).

As for implementation details, would be, one of these containing files/folder regexes of ones to include or exclude from the build process:
- inclusion list or exclusion list
- inclusion list + exclusion list combined
- .gitignore like file

I have implemented something similar with just 10 lines of shell script and it helped separate the two types of builds.

I do agree, no system is 100% foolproof, so there might be once in a while, a PR might slip through or might be flagged incorrectly. But since every PR is also reviewed by human(s) as well we can overcome this small chance.

**Comment by MerelTheisenQB** [comment]
Thanks for the suggestion, but this is beyond the scope of the Kedro project and we're happy with our build process at the moment.
